Feature/ios 3828 add azero blockchain

skibinalexander opened the pull request:
- Добавил интеграцию с Azero
- Пришлось немного изменения инциализации сущностей Polkadot

@megakoko Посмотри пожлалуйста, ибо если не поправить было поведение network получались какие то костыли с прокалыванием параметров

skibinalexander committed:
IOS-3828 Add implementation azero integration

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

skibinalexander committed:
IOS-3828 Add azero test cases

megakoko commented on BlockchainSdk/Extensions/TWCoin+.swift:
```diff
@@ -42,6 +42,8 @@ extension CoinType {
             self = .polkadot
         case .kusama:
             self = .kusama
+        case .azero:
+            return nil
```
Внизу есть отдельный кейс для таких сетей

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -43,6 +50,16 @@ enum PolkadotNetwork: CaseIterable {
+        case .azero:
+            if isTestnet {
+                return [
+                    URL(string: "https://rpc.test.azero.dev")!
+                ]
+            } else {
+                return [
+                    URL(string: "https://rpc.azero.dev")!
```
Нужны запятые в конце

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotWalletManager.swift:
```diff
@@ -160,14 +159,29 @@ extension PolkadotWalletManager: TransactionSender {
-    var existentialDeposit: Amount {
-        network.existentialDeposit
+    
+    func existentialDeposit(blockchain: Blockchain) -> Amount {
```
Есть wallet.blockchain

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotWalletManager.swift:
```diff
@@ -160,14 +159,29 @@ extension PolkadotWalletManager: TransactionSender {
+            return Amount(with: blockchain, value: 0.0000000005)
+        }
     }
+    
```
Лишняя строка

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotWalletManager.swift:
```diff
@@ -160,14 +159,29 @@ extension PolkadotWalletManager: TransactionSender {
+            // This is the lowest amount that activates an account on the Westend network.
+            return Amount(with: blockchain, value: 0.01)
+        case .azero:
+            return Amount(with: blockchain, value: 0.0000000005)
```
Напиши в коде пожалуйста откуда значение взял

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -12,20 +12,27 @@ enum PolkadotNetwork: CaseIterable {
     case polkadot
     case kusama
     case westend
+    case azero
```
Давай зашьем внутрь `testnet` чтобы не переделывать архитектуру

skibinalexander committed:
IOS-3828 Fix refactoring

skibinalexander committed:
IOS-3828 Add faucet link

skibinalexander committed:
IOS-3828 Add azero address test

skibinalexander committed:
IOS-3828 Add draft azero blockchain

skibinalexander committed:
IOS-3828 IOS-2828 Add azero tests transaction

skibinalexander committed:
IOS-3828 Add TODO for polkadot blockchain and tests

skibinalexander committed:
IOS-3828 update parse name blockchain

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -55,14 +85,15 @@ enum PolkadotNetwork: CaseIterable {
             prefixByte = 0
         case .kusama:
             prefixByte = 2
-        case .westend:
+        case .westend, .azero:
             prefixByte = 42
         }
         
         return Data(prefixByte)
     }
-    
-    // https://support.polkadot.network/support/solutions/articles/65000168651-what-is-the-existential-deposit-
+}
+
+extension PolkadotNetwork: ExistentialDepositProvider {
     var existentialDeposit: Amount {
```
Зачем удалил?

skibinalexander replied:
Вернул

megakoko replied:
PolkadotNetwork не нужен протокол ExistentialDepositProvider. Отреверти ненужные изменения пожалуйста. По всему МРу

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -75,6 +106,10 @@ enum PolkadotNetwork: CaseIterable {
             // This value was found experimentally by sending transactions with different values to inactive accounts.
             // This is the lowest amount that activates an account on the Westend network.
             return Amount(with: blockchain, value: 0.01)
+        case .azero:
+            // Existential deposit - 0.0000000005
+            // TODO: - Link on documentation
+            return Amount(with: blockchain, value: 0.0000000005)
```
TODO

megakoko commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotWalletManager.swift:
```diff
@@ -159,15 +158,9 @@ extension PolkadotWalletManager: TransactionSender {
     }
 }
 
-extension PolkadotWalletManager: ExistentialDepositProvider {
-    var existentialDeposit: Amount {
-        network.existentialDeposit
-    }
-}
-
```
Зачем удалил?

skibinalexander replied:
Вернул

megakoko commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -534,23 +534,26 @@ class AddressesTests: XCTestCase {
-    func testSubstrateNetwork(_ network: PolkadotNetwork, publicKey: Data, expectedAddress: String) {
-        let otherNetworks = PolkadotNetwork.allCases.filter { $0 != network }
-        let blockchain = network.blockchain
+    func testAzero() {
+        testSubstrateNetwork(
+            .polkadot(testnet: true),
+            publicKey: edKey,
+            expectedAddress: "5FgMiSJeYLnFGEGonXrcY2ct2Dimod4vnT6h7Ys1Eiue9KxK"
+        )
```
Тут все норм?

skibinalexander replied:
Вынес в общий файл поменял

megakoko commented on BlockchainSdkTests/Polkadot/AzeroTests.swift:
Давай лучше эти тесты сложим в PolkadotTests. Протокол один все равно

skibinalexander commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -1002,6 +1017,7 @@ extension Blockchain {
         case "fantom": return .fantom(testnet: isTestnet)
         case "polkadot": return .polkadot(testnet: isTestnet)
         case "kusama": return .kusama
+        case "aleph zero": return .azero(testnet: isTestnet)
```
Не правильно оттестировал, через - надо спс

skibinalexander committed:
IOS-3828 Add changes after review comments

skibinalexander committed:
IOS-3828 Fix review comments

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

Balashov152 commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotTransactionBuilder.swift:
```diff
@@ -46,9 +46,9 @@ class PolkadotTransactionBuilder {
     private let sigTypeEd25519: UInt8 = 0x00
     
-    init(walletPublicKey: Data, network: PolkadotNetwork) {
+    init(blockchain: Blockchain, walletPublicKey: Data, network: PolkadotNetwork) {
```
а зачем здесь весь `enum Blockchain`?
можно только `PolkadotNetwork` обойтись
а то у нас опять все везде нужно)

Balashov152 replied:
понял что только ради `decimalValue` :/ 
Рефакторить такое очень сложно(
Ладно, потом буду смотреть )

skibinalexander replied:
> понял что только ради `decimalValue` :/
> Рефакторить такое очень сложно(
> Ладно, потом буду смотреть )

Тут просто грань рефакторинга нужно выработать, меня @megakoko заставил откатить изменения )

skibinalexander replied:
> а зачем здесь весь `enum Blockchain`?
> можно только `PolkadotNetwork` обойтись
> а то у нас опять все везде нужно)

Без глубокого рефакторинга не понятно как переиспользовать transaction builder для 3х блокчейнов: polkadot, kusama, azero

Balashov152 replied:
Понял, может у меня руки дойдут )

Balashov152 commented on BlockchainSdkTests/Common/AddressTests.swift:
```diff
@@ -534,23 +534,26 @@ class AddressesTests: XCTestCase {
     
     func testWestend() {
         testSubstrateNetwork(
-            .westend,
```
добавь где нибудь комент плиз, что `westend` это testnet польки

skibinalexander replied:
> добавь где нибудь комент плиз, что `westend` это testnet польки

добавил в PolkadotNetwork

Balashov152 commented on BlockchainSdkTests/Polkadot/PolkadotTests.swift:
```diff
@@ -123,6 +123,43 @@ class PolkadotTests: XCTestCase {
+        let txBuilder = PolkadotTransactionBuilder(blockchain: blockchain, walletPublicKey: publicKey, network: network)
+        
+        let amount = Amount(with: blockchain, value: 12345 / blockchain.decimalValue)
+        let destination = try! blockchain.makeAddresses(from: toAddress, with: nil).first!.value
```
сделай плиз PolkadotAddressService()
blockchain не будет уметь делать адрес

skibinalexander replied:
> сделай плиз PolkadotAddressService()
> blockchain не будет уметь делать адрес

Сделаю ок, но надо обсудить, потому что откатывать изменения приходится :)

megakoko replied:
Это копипаста существующего кода. Давайте не рефакторить сейчас всю польку из-за грядущих изменений. В этом МРе надо только азеро добавить. Остальное отдельно

skibinalexander replied:
Ну конкретно здесь не вижу проблем помочь чтобы дальше чуть проще, в остальных отписался

Balashov152 replied:
не понимаю в чем поинт тесты на новый `Azero` писать не правильно, просто потому-что

megakoko replied:
Чтобы была последовательность, чтобы везде было как везде. Потом все равно надо будет в этот файл идти и править тесты обычной польки.

skibinalexander committed:
IOS-3828 Implement tests via address service

skibinalexander committed:
IOS-3828 Add comments azero blockchain

tureck1y commented on BlockchainSdk/Common/Derivations/DerivationConfigV1.swift:
```diff
@@ -61,6 +61,8 @@ public struct DerivationConfigV1: DerivationConfig {
         case .kusama:
             return [.default: "m/44'/434'/0'/0/0"]
+        case .azero:
+            return [.default: "m/44'/643'/0'/0/0"]
```
Почему именно такой derivationPath? С разработчиками консультировались какой взять? У польки и кусамы путь иначе формируется у траста и других, что отражено в конфиге V3. Azero того же поля ягода. Нужен качественный ресерч, выбрать деривацию правильную и поставить ее одинаковую во все конфиги

skibinalexander replied:
> Почему именно такой derivationPath? С разработчиками консультировались какой взять? У польки и кусамы путь иначе формируется у траста и других, что отражено в конфиге V3. Azero того же поля ягода. Нужен качественный ресерч, выбрать деривацию правильную и поставить ее одинаковую во все конфиги

Возьму в работу

tureck1y replied:
деривацию сделай пожалуйста одинаковую во всех конфигах для этой сети. v1 должен отличаться только для evm сетей

skibinalexander commented on BlockchainSdk/WalletManagers/Polkadot/PolkadotNetwork.swift:
```diff
@@ -8,10 +8,32 @@
+        }
+    }
     
     var blockchain: Blockchain {
```
> Блокчейн вроде не нужен больше тут?

Он нужен вот здесь: 

extension PolkadotNetwork {
    var existentialDeposit: Amount {
        switch self {
        case .polkadot:
            return Amount(with: blockchain, value: 1)
        case .kusama:
            // This value was ALSO found experimentally, just like the one on the Westend.
            // It is different from what official documentation is telling us.
            return Amount(with: blockchain, value: 0.000033333333)
        case .westend:
            // This value was found experimentally by sending transactions with different values to inactive accounts.
            // This is the lowest amount that activates an account on the Westend network.
            return Amount(with: blockchain, value: 0.01)
        case .azero:
            // Existential deposit - 0.0000000005 Look https://test.azero.dev wallet for example
            return Amount(with: blockchain, value: 0.0000000005)
        }
    }
}

Там был обратный мапинг блокчейна, сейчас должно выглядеть более безопасно, когда нам в assembly прилетает нужны

tureck1y replied:
Давай отрефакторим

skibinalexander committed:
IOS-3828 Add azero derivation and fix review comments

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

# Conflicts:
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdkTests/Common/AddressTests.swift
#	BlockchainSdkTests/Polkadot/PolkadotTests.swift

skibinalexander committed:
IOS-3828 Revert tests for polka

tureck1y commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -545,60 +545,7 @@ extension Blockchain {
-            let coin = CoinType(self)!
-            return WalletCoreAddressService(coin: coin, publicKeyType: coin.publicKeyType)
-        }
+        AddressServiceFactory(blockchain: self).makeAddressService().validate(address)
```
а зачем это изменение тут?

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

skibinalexander committed:
IOS-3828 Remove from compare trustwallet azero

skibinalexander committed:
IOS-3828 Merge branch 'develop' into feature/IOS-3828-add-azero-blockchain

skibinalexander committed:
IOS-3828 Add dwellir api rpc

skibinalexander committed:
IOS-3828 revert api dwellir

skibinalexander committed:
IOS-3828 Update derivation for v2 v1 azero
